ambari-server/src/main/java/org/apache/ambari/server/events/publishers/ServiceUpdateEventPublisher.java

```diff
@@ -19,52 +19,62 @@
 package org.apache.ambari.server.events.publishers;
 
 import java.util.ArrayList;
+import java.util.HashMap;
 import java.util.List;
+import java.util.Map;
 
+import org.apache.ambari.server.controller.utilities.ServiceCalculatedStateFactory;
+import org.apache.ambari.server.controller.utilities.state.ServiceCalculatedState;
 import org.apache.ambari.server.events.ServiceUpdateEvent;
+import org.apache.ambari.server.state.State;
 
 import com.google.common.eventbus.EventBus;
 import com.google.inject.Singleton;
 
 @Singleton
 public class ServiceUpdateEventPublisher extends BufferedUpdateEventPublisher<ServiceUpdateEvent> {
+  private Map<String, Map<String, State>> states = new HashMap<>();
 
-  @Override
-  protected Runnable getScheduledPublisher(EventBus m_eventBus) {
-    return new ServiceEventRunnable(m_eventBus);
-  }
-
-  private class ServiceEventRunnable implements Runnable {
 
-    private final EventBus eventBus;
-
-    public ServiceEventRunnable(EventBus eventBus) {
-      this.eventBus = eventBus;
+  @Override
+  public void mergeBufferAndPost(List<ServiceUpdateEvent> events, EventBus eventBus) {
+    List<ServiceUpdateEvent> filtered = new ArrayList<>();
+    for (ServiceUpdateEvent event : events) {
+      int pos = filtered.indexOf(event);
+      if (pos != -1) {
+        if (event.isStateChanged()) {
```

Comment:
Should probably move updates to new method so might be easy to maintain state of event in future.